src/main/java/org/intellij/sdk/codesync/CodeSyncSetup.java

```diff
@@ -0,0 +1,499 @@
+                    return false;
+                }
+
+                // User has access to repo sync, and can proceed.
+                return true;
+            } catch (RequestError e) {
+                // Unknown token status, need to authenticate again.
+                return false;
+            }
+        }
+
+        boolean shouldSignup = Messages.showYesNoDialog(
+                "Do you want to proceed with authentication?",
+                "You Need to Authenticate!",
+                Notification.YES,
+                Notification.NO,
+                Messages.getQuestionIcon()
+        ) == Messages.YES;
+
+        // If user said no to signup request the return here.
+        if (!shouldSignup) {
+            // TODO: We should add logic here to not ask the user again next time he opens the project.
+            return false;
+        }
+
+        CodeSyncAuthServer server;
+        // TODO: Add code to resume repo-sync after successful login. This might be already happening, need to test.
+        try {
+            server =  CodeSyncAuthServer.getInstance();
+            BrowserUtil.browse(server.getAuthorizationUrl());
+        } catch (Exception exc) {
+            exc.printStackTrace();
+            CodeSyncLogger.logEvent(
+                    "[INTELLIJ_AUTH_ERROR]: IntelliJ Login Error, an error occurred during user authentication."
+            );
+            NotificationManager.notifyError("There was a problem with login, please try again later.");
+        }
+
+        return false;
+    }
+
+    /*
+    Start an async task to sync repo.
+     */
+    public static void syncRepoAsync(String repoPath, String repoName, String branchName) {
+        boolean isPublic = Messages.showYesNoDialog(
+                Notification.PUBLIC_OR_PRIVATE,
+                Notification.PUBLIC_OR_PRIVATE,
+                Notification.YES,
+                Notification.NO,
+                Messages.getQuestionIcon()
+        ) == Messages.YES;
+
+        Thread newThread = new Thread(() -> {
+            syncRepo(repoPath, repoName, branchName, isPublic);
+        });
+        newThread.start();
+    }
+
+    public static void syncRepo(String repoPath, String repoName, String branchName, boolean isPublic) {
+        // create .syncignore file.
+        createSyncIgnore(repoPath);
+
+        String[] filePaths = listFiles(repoPath);
+        System.out.println(Arrays.toString(filePaths));
+
+        // Copy files to shadow repo.
+        ShadowRepoManager shadowRepoManager = new ShadowRepoManager(repoPath, branchName);
+        shadowRepoManager.copyFiles(filePaths);
+
+        // Copy files to originals repo.
+        OriginalsRepoManager originalsRepoManager = new OriginalsRepoManager(repoPath, branchName);
+        originalsRepoManager.copyFiles(filePaths);
+
+        // Upload the repo.
+        uploadRepo(repoPath, repoName, filePaths, isPublic);
+
+        // Remove originals repo.
+        originalsRepoManager.delete();
```

Comment:
I think this should only be called once repo is uploaded successfully. In case of no internet, we'll keep the data in .originals repo to try again once internet is restored. I think in JS side, we have this in the finializer of uploadToS3